nexus/src/app/background/tasks/ereport_ingester.rs

```diff
@@ -0,0 +1,680 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Background tasks for ereport ingestion
+
+use crate::app::background::BackgroundTask;
+use anyhow::Context;
+use chrono::Utc;
+use ereport_types::Ena;
+use ereport_types::Ereport;
+use ereport_types::EreportId;
+use futures::future::BoxFuture;
+use internal_dns_types::names::ServiceName;
+use nexus_db_queries::context::OpContext;
+use nexus_db_queries::db;
+use nexus_db_queries::db::DataStore;
+use nexus_types::internal_api::background::EreporterStatus;
+use nexus_types::internal_api::background::SpEreportIngesterStatus;
+use nexus_types::internal_api::background::SpEreporterStatus;
+use omicron_uuid_kinds::EreporterRestartUuid;
+use omicron_uuid_kinds::OmicronZoneUuid;
+use parallel_task_set::ParallelTaskSet;
+use std::collections::BTreeMap;
+use std::net::SocketAddrV6;
+use std::sync::Arc;
+
+pub struct SpEreportIngester {
+    resolver: internal_dns_resolver::Resolver,
+    inner: Ingester,
+}
+
+#[derive(Clone)]
+struct Ingester {
+    datastore: Arc<DataStore>,
+    nexus_id: OmicronZoneUuid,
+}
+
+impl BackgroundTask for SpEreportIngester {
+    fn activate<'a>(
+        &'a mut self,
+        opctx: &'a OpContext,
+    ) -> BoxFuture<'a, serde_json::Value> {
+        Box::pin(async move {
+            let status =
+                self.actually_activate(opctx).await.unwrap_or_else(|error| {
+                    SpEreportIngesterStatus {
+                        error: Some(format!("{error:#}")),
+                        ..Default::default()
+                    }
+                });
+            serde_json::json!(status)
+        })
+    }
+}
+
+impl SpEreportIngester {
+    #[must_use]
+    pub fn new(
+        datastore: Arc<DataStore>,
+        resolver: internal_dns_resolver::Resolver,
+        nexus_id: OmicronZoneUuid,
+    ) -> Self {
+        Self { resolver, inner: Ingester { datastore, nexus_id } }
+    }
+
+    async fn actually_activate(
+        &mut self,
+        opctx: &OpContext,
+    ) -> anyhow::Result<SpEreportIngesterStatus> {
+        // Find MGS clients.
+        // TODO(eliza): reuse the same client across activations; qorb, etc.
+        let mgs_clients = self
+            .resolver
+            .lookup_all_socket_v6(ServiceName::ManagementGatewayService)
+            .await
+            .context("looking up MGS addresses")?
+            .into_iter()
+            .map(|addr| {
+                let url = format!("http://{addr}");
+                let log = opctx.log.new(o!("gateway_url" => url.clone()));
+                let client = gateway_client::Client::new(&url, log);
+                GatewayClient { addr, client }
+            })
+            .collect::<Arc<[_]>>();
+
+        // TODO(eliza): what seems like an appropriate parallelism? should we
+        // just do 16?
```

Comment:
Agreed that this seems a little arbitrary to me.  16 might be fine until we have a compelling reason to tweak things.